### PR TITLE
Deploy Private-Events to production

### PR DESCRIPTION
**Pull request by jubaan**
This pull request merges the private-events app to development branch.
The development was implemented in the following phases, as describe in the commit history:

Set up the rails app
Implementing devise-gem as authentication measure
Creating Models: User(host and attendees) Events and Appointments
Relating the models
Creating scopes and queries
Developing the UX-UI
Updating the README file

### CODE REVIEW

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            <%= link_to('Sign Up', new_user_registration_path, class: "text-decoration-none btn btn-outline-primary ml-1")  %>
+          <% end %>
+        </li>
+        <% if user_signed_in? && controller_name != 'users' %>
+
+          <li class='nav-item'>
+            <button class="btn btn-secondary ml-1">
+              <i class="fa fa-user-circle-o text-white mr-1"></i>
+              <%= link_to(current_user.username, current_user, class: "text-white text-decoration-none")  %>
+            </button>
```
Comment (majovanilla) [inline app/views/layouts/application.html.erb]: This link is broken. When I click it, it takes a while to redirect to the user page but there's an error showing. User show should list all the events the user has created or attended. See [requirements](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png).

File: app/models/event.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+class Event < ApplicationRecord
+  belongs_to :host, foreign_key: 'host_id', class_name: 'User'
+  has_many :appointments, foreign_key: 'event_id', dependent: :destroy
```
Comment (majovanilla) [inline app/models/event.rb]: On line 3 you don't need to specify the foreign key when you include the information in the migration from the beginning. Check more on the migration comment.

**Optional**: You are not validating any attributes on your model. This allows the user to create empty events and that shouldn't happen. You can fix that to improve your project.

File: app/models/event.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+           -> { where('status =?', 'confirmed') },
+           through: :appointments,
+           class_name: 'User',
+           source: :attendee
+
+  has_many :invited_users,
+           -> { where('status =?', 'invited') },
+           through: :appointments,
+           class_name: 'User',
+           source: :attendee
```
Comment (majovanilla) [inline app/models/event.rb:21]: This looks great! Congratulatoins!

File: db/schema.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+# be faster and is potentially less error prone than running all of your
+# migrations from scratch. Old migrations may fail to apply correctly if those
+# migrations use external dependencies or application code.
+#
+# It's strongly recommended that you check this file into your version control system.
+
+ActiveRecord::Schema.define(version: 20_200_616_221_401) do
+  create_table 'appointments', force: :cascade do |t|
+    t.integer 'attendee_id', null: false
+    t.integer 'event_id', null: false
```
Comment (majovanilla) [inline db/schema.rb]: When you create a migration using the following, rails automatically sets the right association, sets the foreign_key option, and knows what class is referencing: `rails g migration nameOfMigration event:references`
That only works when the name of the column is the same as the model you are referencing.

File: app/views/events/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  </div>
+
+  <br>
+
+  <strong>Invite friends</strong>
+  <br>
+
+  <% if @event.invitable_users(current_user).empty? %>
+    <div class="text-muted font-italic">There are no friends to invite</div>
+  <% end %>
```
Comment (majovanilla) [inline app/views/events/show.html.erb:39]: A user that is already attending an event shouldn't appear on this list, but it does. In other words: if the user is already attending, why invite them?

Also, I created two users. One attending some events the other created and for some reason, the user appears as confirmed, pending to confirm and able to invite.

Screenshot from 2020-06-17 22-53-00

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  # GET /users
+  # GET /users.json
+  def index
+    @users = User.all
+  end
+
+  # GET /users/1
+  # GET /users/1.json
+  def show; end
+
```
Comment (majovanilla) [inline app/controllers/users_controller.rb]: As commented  elsewhere, you are not including the information on the User Show page. You should render the events the user's attended and created events.

File: app/views/appointments/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+  <tbody>
+    <% @appointments.each do |appointment| %>
+      <tr>
+        <td><%= appointment.join %></td>
+        <td><%= appointment.leave %></td>
+        <td><%= link_to 'Show', appointment %></td>
+        <td><%= link_to 'Edit', edit_appointment_path(appointment) %></td>
+        <td><%= link_to 'Destroy', appointment, method: :delete, data: { confirm: 'Are you sure?' } %></td>
+      </tr>
```
Comment (majovanilla) [inline app/views/appointments/index.html.erb]: You are not providing a link to access this page. When I go directly on the browser, and I don't have an appointment, I can still see the links and that shouldn't happen.

**Review by majovanilla — CHANGES_REQUESTED** [review]
Hey team, 👋

## You did a *great* job on this project! 💜 

Your page has enough design to be able to navigate through it. Thanks for that! :laughing: 

There are still some changes 👇 you need to implement to approve this request 🔍 

- Please delete the commented code. You have a lot on your controllers.
- About testing. Remember to create your app without the tests so you cand add rspec later. You are missing both integration (authentication and events management) and unit (associations and validations) testing.
- Check the comments below for more details.

### Optional Changes: ✍️
- I left some comments below.
- I strongly suggest you avoid using scaffolds when building your rails app. It gives you a lot of unnecessary elements and it prevents you from building others you must know. For example, you don't need to render both JSON and HTML!


## Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

File: app/views/layouts/application.html.erb (hunk 1 above)
Comment (jubaan) [reply app/views/layouts/application.html.erb]: Done!

File: app/models/event.rb (hunk 2 above)
Comment (jubaan) [reply app/models/event.rb]: Done!

File: app/views/events/show.html.erb (hunk 5 above)
Comment (jubaan) [reply app/views/events/show.html.erb:39]: Done!

File: app/controllers/users_controller.rb (hunk 6 above)
Comment (fc-anjos) [reply app/controllers/users_controller.rb]: Done!

Comment (jubaan) [reply app/controllers/users_controller.rb]: Done

File: app/views/appointments/index.html.erb (hunk 7 above)
Comment (jubaan) [reply app/views/appointments/index.html.erb]: Done!

**Review by wathigo — APPROVED** [review]
### Approved :heavy_check_mark: 
Great work team :clap: 
You have done a great job on this project.

**You may proceed to the next project.**

**Good luck!**
**Simon Wathigo**

**Pull request merged**